### DISCUSSION TITLE
关于训练时使用物品的计算

### DISCUSSION THREAD

**Discussion by robo-dani** [NateScarlet/auto-derby#320] [Ideas]
能不能把体力心情回复 和 训练效率分开计算, 这样复杂度能从 M*N 变成 M+N. 

体力心情可以搞成一个不同档位的费用表, 然后训练根据体力需求, 选择不同档位.
心情就直接默认尽力拉满, 因为道具种类有些 体力表不会复杂,  +5 的体力药 断连战空体的意义更大,  可以当作特例,只用一次之类的(当然为了以后的新道具确实可以都考虑)
苦药汁可能麻烦点, 我打算在心情药溢出时再用,

然后每个训练可以确定一个回体下限和回体上限, 对每个符合下限, 计算上限内的体力回复价值. 然后选择最优之类的.  体力上限可以通过下回合有没有G1 比赛确定.

**Comment by NateScarlet** [comment]
你可以把nurturing.py复制一份自己写自定义逻辑

搞不懂你的目的 要快速直接选别的剧本就完了 要高分用你的方法出来反而降低上限来换速度

**Comment by robo-dani** [comment]
我只是想优化颠峰杯的速度(快速刷历战种马)罢了. 抱歉我一开始确实想当然了.  

不过仔细看了代码后, 我感觉发现plan慢的问题所在了,  
https://github.com/NateScarlet/auto-derby/blob/957f4f8a4817b54b1e9f6065d6e74c0c6e7bf2ee/auto_derby/single_mode/item/item.py#L103-L104
每次算物品分的时候复制两次 context, 脚本跑到后面 race_hisotry, item_history 之类的都变得很大, 导致复制速度变慢了.
我简单的把ctx.__deepcopy__ 改了一下, 碰到history 相关的 就直接引用 https://github.com/robo-dani/auto-derby/commit/413e4a0da3b86f20bfe9e238b4074585c2c51a84

速度快了很多, 基本在0-2s 左右, (改之前最多到了20多, 平均 8s),  不过我只跑了一次, 可能出问题.
想问下pr 前是直接输入 pytest 测试就就可以了吗?

**Reply by NateScarlet** [reply]
直接提 有 github CI

**Reply by NateScarlet** [reply]
我直接把所有 deepcopy 都改成手写了 bf45a5b03a67fcb0fe5ddfb715068908d8db7d17

**Reply by robo-dani** [reply]
呃, 你这样等于重新实现了一次shadow copy了, 不过这样按需deep copy也挺好的。我用copy.copy把重复的代码去掉吧,.

**Reply by NateScarlet** [reply]
我的理念是代码不要太magic 按字符串名称没法用IDE重构

**Reply by NateScarlet** [reply]
之后改一下用 copy.copy 不然用户插件存自定义类上的的数据就丢了
